File: packages/domain/src/validation.ts
import type { DecodeEstimate, DiagnosticsBundle, ProjectFile, ValidationIssue } from "./types";
import { validateGridSpans } from "./grid";

export const validateProject = (project: ProjectFile): ValidationIssue[] => {
  const issues: ValidationIssue[] = [];

  if (project.range.endMs <= project.range.startMs) {
    issues.push({
      level: "error",
      path: "range",
      message: "Range end must be greater than range start.",
    });
  }

  if (project.playback.frameStep < 1) {
    issues.push({
      level: "error",
      path: "playback.frameStep",
      message: "Frame step must be at least 1.",
    });
  }

  if (project.analysisMode === "quick_preview" && project.playback.frameStep > 1) {
    issues.push({
      level: "warning",
      path: "analysisMode",
      message: "Exact frame stepping stays approximate until full fidelity mode is enabled.",
    });
  }

  return issues.concat(validateGridSpans(project.tiles, project.grid));
};

export const estimateFullFidelity = (
  project: ProjectFile,
  fileSizeBytes = 0,
): DecodeEstimate => {
  const width = project.video.width ?? 1920;
  const height = project.video.height ?? 1080;
  const durationSeconds = (project.video.durationMs ?? 60_000) / 1000;
  const fps = project.video.fps ?? 24;
  const projectedMemoryMb = Math.max(32, Math.ceil((width * height * 4) / 1_048_576));
  const projectedCacheMb = Math.max(
    64,
    Math.ceil((durationSeconds * fps) / 8) + Math.ceil(fileSizeBytes / 1_048_576 / 4),
  );
  const canUpgradeToFullFidelity = projectedCacheMb < 2_048;

  return {
    projectedMemoryMb,
    projectedCacheMb,
    canUpgradeToFullFidelity,
    reason: canUpgradeToFullFidelity
      ? undefined
      : "Estimated cache size exceeds the current safety limit.",
  };
};

export const buildDiagnosticsBundle = (project: ProjectFile): DiagnosticsBundle => ({
  appVersion: "0.1.0",
  analysisMode: project.analysisMode,
  videoPath: project.video.path,
  cacheDirectory: "./cache",
  notices: [
    {
      name: "FFmpeg",
      license: "LGPL/GPL depending on distribution",
      url: "https://ffmpeg.org/legal.html",
    },
    {
      name: "Tauri",
      license: "MIT OR Apache-2.0",
      url: "https://v2.tauri.app",
    },
  ],
});

File: crates/vpg-core/src/validation.rs
use crate::grid::validate_grid_spans;
use crate::project::{AnalysisMode, ProjectFile};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ValidationLevel {
    Error,
    Warning,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValidationIssue {
    pub level: ValidationLevel,
    pub path: String,
    pub message: String,
}

/// Validates cross-field project invariants.
pub fn validate_project(project: &ProjectFile) -> Vec<ValidationIssue> {
    let mut issues = Vec::new();

    if project.range.end_ms <= project.range.start_ms {
        issues.push(ValidationIssue {
            level: ValidationLevel::Error,
            path: "range".to_string(),
            message: "range end must be greater than range start".to_string(),
        });
    }

    if project.range.sample_start_ms < project.range.start_ms
        || project.range.sample_start_ms > project.range.end_ms
    {
        issues.push(ValidationIssue {
            level: ValidationLevel::Error,
            path: "range.sampleStartMs".to_string(),
            message: "sample start must stay inside the selected range".to_string(),
        });
    }

    if project.playback.frame_step == 0 {
        issues.push(ValidationIssue {
            level: ValidationLevel::Error,
            path: "playback.frameStep".to_string(),
            message: "frameStep must be at least 1".to_string(),
        });
    }

    if let Some(crop) = project.video.crop {
        if crop.width <= 0.0 || crop.width > 1.0 || crop.height <= 0.0 || crop.height > 1.0 {
            issues.push(ValidationIssue {
                level: ValidationLevel::Error,
                path: "video.crop".to_string(),
                message: "crop width and height must both be greater than 0 and at most 1"
                    .to_string(),
            });
        }

        if crop.x < 0.0 || crop.y < 0.0 || crop.x >= 1.0 || crop.y >= 1.0 {
            issues.push(ValidationIssue {
                level: ValidationLevel::Error,
                path: "video.crop".to_string(),
                message: "crop origin must stay inside the normalized video frame".to_string(),
            });
        }

        if crop.x + crop.width > 1.0 || crop.y + crop.height > 1.0 {
            issues.push(ValidationIssue {
                level: ValidationLevel::Error,
                path: "video.crop".to_string(),
                message: "crop rectangle must fit inside the normalized video frame".to_string(),
            });
        }
    }

    if project.analysis_mode == AnalysisMode::QuickPreview && project.playback.frame_step > 1 {
        issues.push(ValidationIssue {
            level: ValidationLevel::Warning,
            path: "analysisMode".to_string(),
            message: "frame stepping will remain approximate until full fidelity mode is enabled"
                .to_string(),
        });
    }

    issues.extend(
        validate_grid_spans(&project.tiles, project.grid.rows, project.grid.columns)
            .into_iter()
            .map(|issue| ValidationIssue {
                level: ValidationLevel::Error,
                path: format!("tiles.{}", issue.tile_id),
                message: issue.reason,
            }),
    );

    issues
}

#[cfg(test)]
mod tests {
    use crate::project::ProjectFile;

    use super::{validate_project, ValidationLevel};

    #[test]
    fn invalid_range_is_reported() {
        let mut project = ProjectFile::starter("movie.mp4");
        project.range.start_ms = 2_000;
        project.range.end_ms = 1_000;

        let issues = validate_project(&project);
        assert_eq!(issues[0].level, ValidationLevel::Error);
    }

    #[test]
    fn invalid_crop_is_reported() {
        let mut project = ProjectFile::starter("movie.mp4");
        project.video.crop = Some(crate::project::VideoCrop {
            x: 0.75,
            y: 0.2,
            width: 0.5,
            height: 0.5,
        });

        let issues = validate_project(&project);
        assert!(issues.iter().any(|issue| issue.path == "video.crop"));
    }
}
